doc/design/refactor/multi_device.md

```diff
@@ -0,0 +1,37 @@
+# Design Doc: Execute the Program with Multi Thread
+
+## Abstract
+
+This Design Doc propose an approach to make the user-defined Op graph
+running with multi threads, we will use an auto transpiler to convert the user-defined
+Op graph to a multi-thread Op graph, and run `ParallelDo` Op to run the
+multi-thread graph.
+
+## Graph Converter
+<img src="src/multi-threads/single-thread@3x.png" width="300">
+
+After converted:
+
+<img src="src/multi-threads/multi-threads@3x.png" width="1000">
+
+## Implement
```

Comment:
My question is same with today's meeting. How about we have two kinds of thread? say aggregating thread and computing thread? Then we can use the thread mutex and conditional variable without design them in our op level.

Another question if we keep the thread same as your post, then which thread should run the merge op(Block 0)? If the optimization step is very heavy, does the computing gradient thread should wait for it?  

Third question, some `LBFGS` optimize algorithm need more than one pass, how should we distribute the computing thread and optimizing thread? 
LBFGS = > http://pytorch.org/docs/master/optim.html#optimizer-step-closure

Reply:
> My question is same with today's meeting. How about we have two kinds of thread? say aggregating thread and computing thread? Then we can use the thread mutex and conditional variable without design them in our op level

I think `aggregating thread and computing thread ` is a great idea, I'm not sure it will improve performance, because we need to know when the `dW` from all threads were been calculated and then execute `Merge`, we need more context switch, and less threads will execute the computing, but how about adding this optimize to TODO?

> Another question if we keep the thread same as your post, then which thread should run the merge op(Block 0)? If the optimization step is very heavy, does the computing gradient thread should wait for it?

 I think we can improve the performance by executing the optimizer with multi-threads, such as distribute multi parameters `W0, W1, dW0, dW1 ...` and execute the optimizer in different threads. And I will add TODO named `Execute the optimizer with multi-threads.`

> Third question, some LBFGS optimize algorithm need more than one pass, how should we distribute the computing thread and optimizing thread?

The same as above, maybe we can improve the performance for future, but maybe we could execute the optimizer with a single thread.